tnat64.c

```diff
@@ -389,21 +388,36 @@ int getpeername(GETPEERNAME_SIGNATURE)
     {
         struct sockaddr_in6 realpeer;
         socklen_t realpeerlen = sizeof(realpeer);
-        int ret = realgetpeername(__fd, (struct sockaddr *)&realpeer, &realpeerlen);
-        if (ret < 0) {
-            return ret;
-        }
+        memcpy(&realpeer, &sockaddr_st, realpeerlen);
+
         if ((!memcmp(&realpeer.sin6_addr, &ipv4mapped, NAT64PREFIXLEN)) || (check_prefix(config, &realpeer.sin6_addr)))
         {
-            struct sockaddr_in * result;
-            result = (struct sockaddr_in *)__addr;
-            result->sin_family = AF_INET;
-            result->sin_port = realpeer.sin6_port;
-            memcpy(&result->sin_addr, &realpeer.sin6_addr.s6_addr[12], sizeof(struct in_addr));
+            struct sockaddr_in result;
+            memset(&result, 0, sizeof(result));
+
+            result.sin_family = AF_INET;
+            result.sin_port = realpeer.sin6_port;
+            memcpy(&result.sin_addr, &realpeer.sin6_addr.s6_addr[12], sizeof(struct in_addr));
+
+            // Copy up to *__len bytes into the available space.
+            memcpy(__addr, &result, *__len);
+
             *__len = sizeof(struct sockaddr_in);
             return ret;
         }
+        else {
+            // Not sure what the best data to return here would be. 
+            // This should never happen in normal operation, though. 
+            // This would only be executed if the socket was connected to an IPv6 address 
+            // that's not part of a NAT64 prefix.
```

Comment:
I guess, at least some
`show_msg(MSGERROR, "<...>");`

Reply:
OTOH probably not, since even if this happens, anything should transparently work as before.

Reply:
I mean, yeah, everything should continue like with older versions of tnat64, but it's still an indicator that something's amiss. But same as with your other comment, I guess this can happen when an application opens IPv6 sockets natively?

Reply:
Added a log warning to the code if that happens.